[DI - Step 2] 에단 미션 제출합니다.

Laterality opened the pull request:
안녕하세요 쏘쏘, 에단입니다!

2단계 미션 제출합니다, 잘 부탁드립니다 👍

Laterality committed:
feat: Configuration Bean Initialization

Laterality committed:
feat: 2단계 요구사항 구현

- 빈 생성을 위한 스캐너 분리 및 추상화.
- ComponentScan 어노테이션 적용.
- ApplicationContext 구현.

Laterality committed:
feat: 2단계 요구사항 구현

- 빈 스캐너 분리 및 추상화.
- ComponentScan 어노테이션 적용.
- ApplicationContext 구현.

Laterality committed:
Merge remote-tracking branch 'origin/step02' into step02

Laterality committed:
chore: 테스트 리소스 추가

ssosso reviewed (changes requested):
흠잡을때가 없네요 에단 💯 
리펙토링 힌트에서 제공되는 빈 등록에 대한 테스트 정도만 추가되면
이번 미션을 끝으로 보이네요. 테스트 추가 부탁드립니다:)

Laterality committed:
Add test for configuration scanner

Laterality committed:
Refactor bean factory

- Changed to take vararg instead of list of bean scanners as constructor argument.

Laterality commented:
피드백 감사합니다. 리팩토링 힌트 참고해서 빈 스캐너에 관한 테스트 추가해봤습니다!

힌트의 테스트 코드에서는 컴포넌트 스캐너가 빈 팩토리를 참조하는 것으로 보이는데요. 저는 반대로 빈 팩토리가 빈 스캐너를 추상화한 인터페이스를 참조하도록 구현했는데, 두 방식에 큰 차이점이 있을까요?
 
스캐너가 팩토리에 의존하면 팩토리의 변경에 영향을 받을 수 있을 거라고 생각했고, 빈 팩토리가 스캐너의 인터페이스를 참조하도록 구현하면 구현체(스캐너)의 변경에 빈 팩토리가 받는 영향이 적거나 없을 거라고 생각해서 지금의 방식으로 구현했는데요. 혹시 제가 놓친 부분이 있는지 궁금합니다😁

ssosso reviewed (approved):
안녕하세요, 에단!
질문주신 의존성의 반대방향은 현재 구조에서는 크게 영향이 없어 보이네요:)
깔끔하게 짜주셔서 더 피드백할 내용이 없네요. 머지하겠습니다~!
